near-sdk-macros/src/core_impl/abi/abi_visitor.rs

```diff
@@ -0,0 +1,75 @@
+use super::TypeRegistry;
+use crate::ItemImplInfo;
+use proc_macro2::TokenStream as TokenStream2;
+use quote::{quote, ToTokens};
+use syn::visit::Visit;
+use syn::{Error, ItemImpl};
+
+/// Information relevant to ABI extracted from the `impl` section decorated with `#[near_bindgen]`.
+#[derive(Default)]
+pub struct AbiVisitor {
+    impl_item_infos: Vec<ItemImplInfo>,
+    /// Errors that occured while extracting the data.
+    errors: Vec<Error>,
+}
+
+impl<'ast> Visit<'ast> for AbiVisitor {
+    fn visit_item_impl(&mut self, i: &'ast ItemImpl) {
+        let has_near_sdk_attr = i
+            .attrs
+            .iter()
+            .any(|attr| attr.path.to_token_stream().to_string().as_str() == "near_bindgen");
+        if has_near_sdk_attr {
+            match ItemImplInfo::new(&mut i.clone()) {
+                Ok(info) => self.impl_item_infos.push(info),
+                Err(err) => self.errors.push(err),
+            }
+        }
+        syn::visit::visit_item_impl(self, i);
+    }
+}
+impl AbiVisitor {
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    pub fn generate_abi_function(&self) -> syn::Result<TokenStream2> {
+        let mut registry = TypeRegistry::new();
+        if !self.errors.is_empty() {
+            return Err(self.errors[0].clone());
+        }
+        let functions: Vec<TokenStream2> = self
+            .impl_item_infos
+            .iter()
+            .flat_map(|i| &i.methods)
+            .map(|m| m.abi_struct(&mut registry))
+            .collect();
+        let types: Vec<TokenStream2> = registry
+            .types
+            .iter()
+            .map(|(t, id)| {
+                quote! {
+                    near_sdk::__private::AbiType { id: #id, schema: gen.subschema_for::<#t>() }
+                }
+            })
+            .collect();
+        Ok(quote! {
+            const _: () = {
+                #[no_mangle]
+                #[cfg(not(target_arch = "wasm32"))]
+                pub fn __near_abi() -> near_sdk::__private::AbiRoot {
```

Comment:
Note that this will fail to compile when there are multiple impl sections with `near_bindgen`, right? Will double check tomorrow

Reply:
Hmm, it should not as long as all impl sections are wrapped in the same `near_abi!` call.

Reply:
didn't realize that case is handled. It limits the ability to have impl sections in diff modules, though

Reply:
True. I am unsure how to handle this though... I have tried inlining modules like this:

```rust
#[near_abi]
mod abi {
    #[path = "utility.rs"]
    mod abi_utility;

    #[near_bindgen]
    #[derive(Default, BorshDeserialize, BorshSerialize)]
    pub struct Contract {}

    ...
}
```

But apparently including non-inline modules in proc macro input is an unstable feature.

I was thinking I could maybe create a global state that would be modified by all `#[near_abi]` annotations and then add another macro that would output the accumulated state into one ABI structure. The problem is that I think there is no way to enforce macro execution order. Do you have any ideas I could explore?

Reply:
Another very vague idea I have considered is to generate a bunch of symbols (e.g. `__near_abi_lib_rs`, `__near_abi_utility_rs` etc) and then somehow get a list of them from cargo extension and call all of them. The main problem here is that I cannot find if there is a rustc/cargo API that would allow me to list all symbols from a crate.